perf(transformer/tagged-template-transform): improve performance

Dunqing opened the pull request:
Made some improvements based on the suggestions in https://github.com/oxc-project/oxc/pull/15664

Dunqing commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.com/github/pr/oxc-project/oxc/15834?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#15834** <a href="https://app.graphite.com/github/pr/oxc-project/oxc/15834?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.com/github/pr/oxc-project/oxc/15834?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#15830** <a href="https://app.graphite.com/github/pr/oxc-project/oxc/15830?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.com/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=15834).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

overlookmotel reviewed (commented):
I think to avoid a perf hit in `enter_expression`, you want it to be:

```rs
impl<'a> Traverse<'a, TransformState<'a>> for TaggedTemplateTransform<'a, '_> {
    // `#[inline]` because this is a hot path and most `Expression`s are not `TaggedTemplateExpression`s,
    // so we want this inlined to handle the common case without a function call
    #[inline]
    fn enter_expression(&mut self, expr: &mut Expression<'a>, ctx: &mut TraverseCtx<'a>) {
        if matches!(expr, Expression::TaggedTemplateExpression(_)) {
          self.transform_tagged_template(expr, ctx);
        }
    }
}
```

The idea is that probably 0.01% of `Expression`s are `TaggedTemplateExpression`s, so for the common case all you want in the hot path is "Is this a `TaggedTemplateExpression`? No it's not. Exit." That's a very small function, so it can be inlined, which means no overhead of a function call on the hot path.

Dunqing committed:
perf(transformer/tagged-template-transform): improve performance (#15834)

Made some improvements based on the suggestions in https://github.com/oxc-project/oxc/pull/15664
